src/test/compile-fail/enum-and-module-in-same-scope.rs

```diff
@@ -13,7 +13,7 @@ mod Foo {
 }
 
 enum Foo {  //~ ERROR duplicate definition of type or module `Foo`
-    X //~ ERROR duplicate definition of value `X`
+    X
```

Comment:
> we're not "merging" the enum/normal modules in some way that we used to do

Yeah, that's exactly what is happening.
When visiting enums, we used to call `NameBindings::set_module_kind`. Assuming duplicate checking passed, `set_module_kind` is the same as `define_module`, but if duplicate checking failed and the type namespace is already a module, `set_module_kind` sets the `kind` of the module but leaves everything else, including the `def_id`.

Setting only the `kind` seemed wrong, especially after refactoring `kind` and `def_id` into `def`, so I replaced all calls to `set_module_kind` with calls to `define_module` after checking that it would not change anything when there are no duplicate errors.

I could recover the original behavior on this test by changing `define_module` to not do anything when the type namespace is already defined.